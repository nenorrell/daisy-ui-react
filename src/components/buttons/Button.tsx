import { FC, ForwardedRef, forwardRef, MouseEventHandler, PropsWithChildren } from "react";
import { BaselessVariant } from "../../@types/Colors";
import { ButtonSize } from "../../@types/Daisy";
import clsx from "clsx";

interface IButton {
    id ?:string
    className ?:string
    size ?:ButtonSize
    variant ?:BaselessVariant
    onClick ?:MouseEventHandler<HTMLButtonElement>
    isGhost ?:boolean
    isLink ?:boolean
    isOutline ?:boolean
    isDisabled ?:boolean
    isGlass ?:boolean
    isLoading ?:boolean
    noAnimation ?:boolean
    isActive ?:boolean
    shape ?:"circle" | "square"
}
export const Button :FC<PropsWithChildren<IButton>> = forwardRef((
    {variant="neutral", size="md", ...props},
    ref ?:ForwardedRef<HTMLButtonElement>
)=>{
    return (
        <button
            id={props.id}
            ref={ref}
            onClick={props.onClick}
            className={clsx(
                "btn",
                {
                    [`btn-${variant}`]: variant,
                    [`btn-${size}`]: size
                },
                props.className,
                props.isGhost && "btn-ghost",
                props.isActive && "btn-active",
                props.isGlass && "glass",
                props.isLink && "btn-link",
                props.isLoading && "loading",
                props.isOutline && "btn-outline",
                props.noAnimation && "no-animation",
                props.shape === "circle" && "btn-circle" ||
                props.shape === "square" && "btn-square"
            )}
            disabled={props.isDisabled}
        >
            {props.children}
        </button>
    );
});

Button.displayName = "Button";
