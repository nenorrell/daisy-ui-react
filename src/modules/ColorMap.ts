import { Variant, VariantColors } from "../@types/Daisy";


interface ColorMap<K, V> extends Map<K, V> {
    get(key: K): V;
}

// Cast to ColorMap interface to get rid of inherit "possibly undefined value" from Map
export default <ColorMap<Variant, VariantColors>>new Map([
    ["primary", {
        bg: "bg-primary",
        text: "text-primary-content",
        focus: "primary-focus",
        btn: "btn-primary"
    }
    ],
    ["secondary", {
        bg: "bg-secondary",
        text: "text-secondary-content",
        focus: "secondary-focus",
        btn: "btn-secondary"
    }
    ],
    ["accent", {
        bg: "bg-accent",
        text: "text-accent-content",
        focus: "accent-focus",
        btn: "btn-accent"
    }
    ],
    ["neutral", {
        bg: "bg-neutral",
        text: "text-neutral-content",
        focus: "neutral-focus",
        btn: "btn-neutral"
    }
    ],
    ["info", {
        bg: "bg-info",
        text: "text-info-content",
        focus: "info-focus",
        btn: "btn-info"
    }
    ],
    ["success", {
        bg: "bg-success",
        text: "text-success-content",
        focus: "success-focus",
        btn: "btn-success"
    }
    ],
    ["error", {
        bg: "bg-error",
        text: "text-error-content",
        focus: "error-focus",
        btn: "btn-error"
    }
    ],
    ["warning", {
        bg: "bg-warning",
        text: "text-warning-content",
        focus: "warning-focus",
        btn: "btn-warning"
    }
    ],
    ["base-100", { bg: "bg-base-100", text: "text-base-content" }],
    ["base-200", { bg: "bg-base-200", text: "text-base-content" }],
    ["base-300", { bg: "bg-base-300", text: "text-base-content" }]
]);
