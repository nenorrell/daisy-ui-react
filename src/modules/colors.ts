import { Variant } from "../@types/Colors";

export const getTextColor = (variant :Variant) =>{
    if(variant.includes("base")) {
        return "base";
    }
    return variant;
};
