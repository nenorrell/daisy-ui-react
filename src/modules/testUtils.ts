import { BaselessVariant, ButtonSize, Variant } from "../@types/Daisy";

export const BaselessVariantOptions :BaselessVariant[] = ["primary", "secondary", "accent", "neutral", "info", "success", "error", "warning"];
export const VariantOptions :Variant[] = [...BaselessVariantOptions, "base-100", "base-200", "base-300"];

export const ButtonSizes :ButtonSize[] = ["lg", "md", "sm", "xs"];
