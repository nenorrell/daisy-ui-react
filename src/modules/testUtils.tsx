import { BaselessVariant, StatusVariant, Variant } from "../@types/Colors";
import { ButtonSize, Size } from "../@types/Daisy";

export const BaselessVariantOptions :BaselessVariant[] = ["primary", "secondary", "accent", "neutral", "info", "success", "error", "warning"];
export const StatusVariantOptions :StatusVariant[] = ["info", "success", "error", "warning"];
export const VariantOptions :Variant[] = [...BaselessVariantOptions, ...StatusVariantOptions, "base-100", "base-200", "base-300"];

export const ComponentSizes :Size[] = ["lg", "md", "sm", "xs"];
export const ButtonSizes :ButtonSize[] = [...ComponentSizes, "block", "wide"];
