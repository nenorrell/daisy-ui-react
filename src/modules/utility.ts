import React, { FunctionComponent } from "react";
import { ComponentOrJSX } from "../@types/Generic";
import ReactIs from "react-is";

export function processComponentOrJSX<P=any>(target :ComponentOrJSX<P>) :FunctionComponent<P> | null {
    if(ReactIs.isValidElementType(target)) {
        return target as React.FunctionComponent<P>;
    }
    const wrapper :FunctionComponent<P> = (props :any) => {
        return React.cloneElement(target as JSX.Element, props);
    };
    return wrapper;
};
