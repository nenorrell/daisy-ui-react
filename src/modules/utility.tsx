import React, { ReactElement, ReactNode } from "react";

export const passPropToChildren = <T=any>(children :ReactElement<T>[] | ReactElement<T>, props :object)=>{
    return React.Children.map<ReactNode, ReactElement>(children, (child) =>
        React.cloneElement(child, {
            ...child.props,
            ...props
        })
    );
};
