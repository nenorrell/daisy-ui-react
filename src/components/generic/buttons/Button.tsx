import { MouseEventHandler } from "react";
import { BaselessVariant, ButtonSize } from "../../../@types/Daisy";
import ColorMap from "../../../modules/ColorMap";
import { getButtonSize } from "../../../modules/utility";

interface IButton {
    buttonText ?:string
    size ?:ButtonSize
    /** Use className for modifiers like btn-ghost, btn-circle, etc */
    className ?:string
    variant ?:BaselessVariant
    onClick ?:MouseEventHandler
}
export const Button = ({className, variant="neutral", size, buttonText, onClick, children}:React.PropsWithChildren<IButton>)=>{
    const variantColors = ColorMap.get(variant);
    const btnSize = getButtonSize(size);

    return (
        <button onClick={onClick ? onClick : undefined} className={`btn ${btnSize} ${variantColors?.btn} ${className ? className : ""}`}>
            {buttonText || children}
        </button>
    );
};
