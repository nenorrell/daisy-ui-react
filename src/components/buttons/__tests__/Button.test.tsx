import { fireEvent, render } from "@testing-library/react";
import ColorMap from "../../../modules/ColorMap";
import { BaselessVariantOptions, ButtonSizes } from "../../../modules/testUtils";
import { getButtonSize } from "../../../modules/utility";
import { Button } from "../Button";

describe("Button", ()=>{
    describe("Variants", ()=>{
        BaselessVariantOptions.forEach(variant=>{
            it(`Respects ${variant} variant`, ()=>{
                const component = render(<Button buttonText={`testing ${variant} variant`} variant={variant} />);
                const container = component.container.querySelector("button.btn");
                const colors = ColorMap.get(variant);

                expect(container?.classList.contains(colors.btn as any)).toBe(true);
                expect(container).toMatchSnapshot();
            });
        });
    });

    describe("Sizes", ()=>{
        ButtonSizes.forEach(size=>{
            it(`Respects ${size} size`, ()=>{
                const component = render(<Button buttonText={`testing ${size} button size`} size={size} />);
                const container = component.container.querySelector("button.btn");
                const sizing = getButtonSize(size);

                expect(container?.classList.contains(sizing)).toBe(true);
                expect(container).toMatchSnapshot();
            });
        });
    });

    it("Respects onClick when passed in", ()=>{
        const stub = jest.fn();
        const component = render(
            <Button
                buttonText="testing"
                onClick={stub}
            />
        );
        const container = component.container.querySelector("button.btn");
        fireEvent.click(container as HTMLButtonElement);
        expect(stub).toHaveBeenCalled();
    });

    it("Disables button when passed in", ()=>{
        const component = render(
            <Button
                buttonText="testing"
                isDisabled={true}
            />
        );
        const container = component.container.querySelector("button.btn");
        expect(container).toBeDisabled();
    });

    it("Does not disable button when disabled is false", ()=>{
        const component = render(
            <Button
                buttonText="testing"
                isDisabled={false}
            />
        );
        const container = component.container.querySelector("button.btn");
        expect(container).not.toBeDisabled();
    });

    it("Renders buttonText instead of children when both are present", ()=>{
        const component = render(
            <Button
                buttonText="testing"
            ><p className="find-me">Some Child</p></Button>
        );
        const container = component.container.querySelector("button.btn");
        expect(container?.firstChild?.nodeValue).toEqual("testing");

        const child = component.container.querySelector("p.find-me");
        expect(child).not.toBeInTheDocument();
    });

    it("Renders children", ()=>{
        const component = render(
            <Button><p className="find-me">Some Child</p></Button>
        );
        const container = component.container.querySelector("p.find-me");
        expect(container).toBeInTheDocument();
    });

    it("applies ghost class", () => {
        const { container } = render(<Button isGhost={true} />);
        const button = container.firstChild;

        expect(button).toHaveClass("btn-ghost");
    });

    it("applies link class", () => {
        const { container } = render(<Button isLink={true} />);
        const button = container.firstChild;

        expect(button).toHaveClass("btn-link");
    });

    it("applies outline class", () => {
        const { container } = render(<Button isOutline={true} />);
        const button = container.firstChild;

        expect(button).toHaveClass("btn-outline");
    });

    it("applies glass class", () => {
        const { container } = render(<Button isGlass={true} />);
        const button = container.firstChild;

        expect(button).toHaveClass("glass");
    });

    it("applies loading class", () => {
        const { container } = render(<Button isLoading={true} />);
        const button = container.firstChild;

        expect(button).toHaveClass("loading");
    });

    it("applies no-animation class", () => {
        const { container } = render(<Button noAnimation={true} />);
        const button = container.firstChild;

        expect(button).toHaveClass("no-animation");
    });

    it("applies circle shape class", () => {
        const { container } = render(<Button shape="circle" />);
        const button = container.firstChild;

        expect(button).toHaveClass("btn-circle");
    });

    it("applies square shape class", () => {
        const { container } = render(<Button shape="square" />);
        const button = container.firstChild;

        expect(button).toHaveClass("btn-square");
    });
});
