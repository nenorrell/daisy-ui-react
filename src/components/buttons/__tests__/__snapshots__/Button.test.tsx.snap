// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Sizes Respects block size 1`] = `
<button
  class="btn btn-neutral btn-block"
>
  testing 
  block
   button size
</button>
`;

exports[`Button Sizes Respects lg size 1`] = `
<button
  class="btn btn-neutral btn-lg"
>
  testing 
  lg
   button size
</button>
`;

exports[`Button Sizes Respects md size 1`] = `
<button
  class="btn btn-neutral btn-md"
>
  testing 
  md
   button size
</button>
`;

exports[`Button Sizes Respects sm size 1`] = `
<button
  class="btn btn-neutral btn-sm"
>
  testing 
  sm
   button size
</button>
`;

exports[`Button Sizes Respects wide size 1`] = `
<button
  class="btn btn-neutral btn-wide"
>
  testing 
  wide
   button size
</button>
`;

exports[`Button Sizes Respects xs size 1`] = `
<button
  class="btn btn-neutral btn-xs"
>
  testing 
  xs
   button size
</button>
`;

exports[`Button Variants Respects accent variant 1`] = `
<button
  class="btn btn-accent btn-md"
>
  testing 
  accent
   variant
</button>
`;

exports[`Button Variants Respects error variant 1`] = `
<button
  class="btn btn-error btn-md"
>
  testing 
  error
   variant
</button>
`;

exports[`Button Variants Respects info variant 1`] = `
<button
  class="btn btn-info btn-md"
>
  testing 
  info
   variant
</button>
`;

exports[`Button Variants Respects neutral variant 1`] = `
<button
  class="btn btn-neutral btn-md"
>
  testing 
  neutral
   variant
</button>
`;

exports[`Button Variants Respects primary variant 1`] = `
<button
  class="btn btn-primary btn-md"
>
  testing 
  primary
   variant
</button>
`;

exports[`Button Variants Respects secondary variant 1`] = `
<button
  class="btn btn-secondary btn-md"
>
  testing 
  secondary
   variant
</button>
`;

exports[`Button Variants Respects success variant 1`] = `
<button
  class="btn btn-success btn-md"
>
  testing 
  success
   variant
</button>
`;

exports[`Button Variants Respects warning variant 1`] = `
<button
  class="btn btn-warning btn-md"
>
  testing 
  warning
   variant
</button>
`;
